### ISSUE TITLE
Request for compute: anyInvocation() and allInvocation()

### ISSUE THREAD

**Issue by krogovin** [gpuweb/gpuweb#2137]
PR's https://github.com/gpuweb/gpuweb/pull/954 and https://github.com/gpuweb/gpuweb/pull/1459 are super-sets of this request. However, this smaller request is to have available in a compute shader the functions `anyInvocation()` and `allInvocation()` in compute shader. For certain cases (for example mine) these make a big difference for performance, especially to prevent divergence within an SIMD group.

**Comment by litherum** [comment]
@mehmetoguzderin did a great investigation of the performance benefits of these operations [here](https://github.com/gpuweb/gpuweb/pull/954#issuecomment-690216791). It's clear they're useful and they do improve performance.

However, they are non-portable, as demonstrated [here](https://github.com/gpuweb/gpuweb/pull/954#issuecomment-693588589), and we have evidence that real authors are hitting the non-portability problems with these sub group operations. Therefore, until they are made to be portable in the native APIs, the CG has decided to postpone this feature until after MVP/v1.

**Comment by krogovin** [comment]
Um, `anyInvocation()` and `allInvocation()` are portable. Indeed a dummy implementation of both `anyInvocation(foo)` and `allInvocation(foo)` is to return the value of `foo` (i.e. pretend that the SIMD width is one). The point is that the underlying API's (atleast Metal and Vulkan) have these. The point is that (for my use case) they are really needed for performance.

**Comment by dneto0** [comment]
There is a distinction between "has an implementation" vs. "the behaviour is consistent across implementations".

Even within a single device, it's not clear when reconvergence occurs. (See the discussion above @litherum's table in #954 ).  Programmers assume things that implementations don't necessarily give them, which leads unhappiness and finger-pointing.  The group's choice to emphasize consistent behaviour means subgroups are out for MVP/v1.

Somewhat related: an investigation into what happens when you sample in non-uniform control flow: [blogpost](https://themaister.net/blog/2019/09/12/the-weird-world-of-shader-divergence-and-lod/)

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-09-28-Minutes#-request-for-compute-anyinvocation-and-allinvocation-2137-">WGSL meeting minutes 2021-09-28</a></summary>



* MM: These functions are no better or works than the other functions that we kicked out of MVP.  No reason to take them in.
* MOD: These also depend on the active mask. So same issues plague it as the whole subgroup feature discussed earlier.
* JG: Can **postpone**.
* GR: Is Invocation a wave or warp or quad?
* MM: A thread
* GR: +1 for postpone then
* MM: This is a good feature that we want, we just have more work to do here before we can ship it.

 


</details>

**Comment by krogovin** [comment]
I am not asking for the case of fragment shading; I explicitly stated the request for **_compute only_** which means a great deal of subtly goes away. The only guarantee that these functions provide is that:
- `anyInvocation(foo)` is `true` whenever `foo` is `true`
- `allInvocations(foo)` is `false` whenever `foo` is `false`

I am **_not_** advocating that code under `anyInvocation(foo)` or `allInvocations(foo)` becomes dynamically uniform. I am perfectly happy with that in terms of control-flow analysis with respect to being dynamically uniform that conditions from `anyInvocation(foo)` or `allInvocations(foo)` are just as bad as directly from `foo`.

The expectation is that a shader using these understand them as essentially a hint on what how the GPU is running its SIMD jazz. All of the targeted API's have these functions, so it really should be there. My use case this is a pretty big freaking deal where I have a compute shader that is walking a list of stuff to do. If it can compute that it can skip a block of work, it would then fast forward. However, if there are work-items in the same SIMD group that do not fast forward, then the SIMD is forever out sync after a fast forward. On the other hand, with the above, I would do fast forward when allInvocations(foo)` is `true`. This makes a killer difference for performance and is essential. From the point of view of what is going on with an SIMD GPU, it is clear as to what it is doing.

As far as discluding this for the sake of having the same behaviour across all platforms: that is suspicious in my eyes. Mipmap computation already has leeway between API's and GPU implementations of it.

**Comment by dneto0** [comment]
This is landed as part of the subgroups feature as `subgroupAny(b:bool)` and `subgroupAll(b:bool)`
